### ISSUE TITLE
The two-phase transaction is committed in master but rollbacked in segments.

### ISSUE THREAD

**Issue by hidva** [greenplum-db/gpdb#12199]
### Greenplum version or build

 PostgreSQL 12beta2 (Greenplum Database 7.0.0-alpha.0+dev.14704.gb1c8b9f108 build dev) on x86_64-pc-linux-gnu, compiled by gcc (GCC) 6.5.1 20190307 (Alibaba 6.5.1-1 2.17), 64-bit compiled on Jun 22 2021 22:54:52 (with assert checking)

### Actual behavior

```
tmp=# create table twopcbug(i int, j int);
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'i' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
CREATE TABLE
tmp=# select * from twopcbug;
ERROR:  could not open relation with OID 24587  (seg0 slice1 172.17.0.3:7002 pid=52886)
DETAIL:  This can be validly caused by a concurrent delete operation on this object.
```

### Step to reproduce the behavior

1.  Create a breakpoint at insertedDistributedCommitted

    ```
    tmp=# select pg_backend_pid();
    51964
    tmp=# begin;
    BEGIN
    tmp=# create table twopcbug(i int, j int);
    NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'i' as the Greenplum Database data distribution key for this table.
    HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
    CREATE TABLE
    ```

    ```
    gdb -p 51964
    (gdb) b insertedDistributedCommitted
    Breakpoint 1 at 0x1c35e14: file cdbtm.c, line 1489.
    (gdb) c
    Continuing.
    ```

    ```
    tmp=# select pg_backend_pid();
    51964
    tmp=# commit;
    ```

    ```
    Breakpoint 1, insertedDistributedCommitted () at cdbtm.c:1489
    1489		ereport(DTM_DEBUG5,
    ```

2.  Create a breakpoint at getDtxCheckPointInfo() in master checkpointer process. `getDtxCheckPointInfo()` got the **empty** committed two-phase transaction because `insertedDistributedCommitted()` in process 51964 is not executed yet.

    ```
    (gdb) b getDtxCheckPointInfo
    Breakpoint 2 at 0x1587079: file procarray.c, line 1884.
    (gdb) c
    Continuing.

    Breakpoint 2, getDtxCheckPointInfo (result=0x7ffee0fdda50, result_size=0x7ffee0fdd990) at procarray.c:1884
    1884		ProcArrayStruct *arrayP = procArray;
    (gdb) finish
    Run till exit from #0  getDtxCheckPointInfo (result=0x7ffee0fdda50, result_size=0x7ffee0fdd990) at procarray.c:1884
    CreateCheckPoint (flags=256) at xlog.c:9099
    9099		LWLockAcquire(XidGenLock, LW_SHARED);
    (gdb) p dtxCheckPointInfo
    $1 = 0x625000016b68 ""
    ```

3.  `killall -9 postgres` after CreateCheckPoint() is finished; 

    ```
    gdb -p 51964
    (gdb) p MyPgXact->delayChkpt
    $1 = true
    (gdb) watch MyPgXact->delayChkpt
    Hardware watchpoint 2: MyPgXact->delayChkpt
    (gdb) c
    Continuing.
    Hardware watchpoint 2: MyPgXact->delayChkpt

    Old value = true
    New value = false
    RecordTransactionCommit () at xact.c:1759
    1759			END_CRIT_SECTION();
    (gdb) quit
    Quitting: Couldn't get registers: No such process.    
    ```

4.  `gpstart -a`; And gxid 8209 is committed in master but rollbacked in segments. 

    ```
    $ pg_waldump ${qddir}/pg_wal/000000010000000000000003
    rmgr: Transaction len (rec/tot):    501/   501, tx:        569, lsn: 0/0C24B338, prev 0/0C24B300, desc: DISTRIBUTED_COMMIT distributed commit 2021-06-22 23:02:07.182021 CST gxid = 8209
    rmgr: XLOG        len (rec/tot):    130/   130, tx:          0, lsn: 0/0C24B600, prev 0/0C24B5C8, desc: CHECKPOINT_ONLINE redo 0/C24B598; tli 1; prev tli 1; fpw true; xid 0:570; gxid 16401; oid 32779; relfilenode 32774; multi 1; offset 0; oldest xid 520 in DB 1; oldest multi 1 in DB 1; oldest/newest commit timestamp xid: 0/0; oldest running xid 569; online, checkpoint record data length = 104, DTX committed count 0, DTX data length 8
    ```

    ```
    $ pg_waldump ${seg0}/pg_wal/000000030000000000000005
    rmgr: Transaction len (rec/tot):   1037/  1037, tx:        568, lsn: 0/15766620, prev 0/157665F0, desc: PREPARE at = 2021-06-22 23:02:07.181167 CST; gid = 8209
    rmgr: Transaction len (rec/tot):     74/    74, tx:          0, lsn: 0/15766B20, prev 0/15766A98, desc: ABORT_PREPARED 568: 2021-06-22 23:07:31.143694 CST; rels: base/16384/32777
    ```

### Root Cause

`getDtxCheckPointInfo()` in checkpointer process is invoked before `insertedDistributedCommitted()` in backends.

**Comment by ashwinstar** [comment]
Thank you for report and Great investigation @hidva.

Basically, `getDtxCheckPointInfo()` should be called after this block of code in `CreateCheckPoint()` as `insertedDistributedCommitted()` is relying on `delayCkpt` to prevent this race condition. Thoughts?

```C
    vxids = GetVirtualXIDsDelayingChkpt(&nvxids); 
    if (nvxids > 0)
    {
        do
        {
            pg_usleep(10000L);  /* wait for 10 msec */
        } while (HaveVirtualXIDsDelayingChkpt(vxids, nvxids));
    }
    pfree(vxids);
```
Would you like to create isolation2 test for this and open PR to fix it?

**Comment by ashwinstar** [comment]
Just for reference, same problem exist on GPDB6 as well, hence the fix will be required to be back-ported to GPDB6 as well.

**Comment by asimrp** [comment]
@hidva cool find!  Thank you for the report as well as the fix.  I'm very curious on how did you find this issue?  Can you please share the trick?

**Comment by hidva** [comment]
> @hidva cool find! Thank you for the report as well as the fix. I'm very curious on how did you find this issue? Can you please share the trick?

I would like to know, what will happen if an XLOG_XACT_DISTRIBUTED_COMMIT record associated with a distributed transaction is inserted before a checkpoint record and we do crash recover from the checkpoint record? Will the distributed transaction be rollbacked? So I read the relevant code and keenly discovered that this is indeed possible. 

I have reimplemented the PG transaction system in [KuiBaDB](https://github.com/KuiBaDB/KuiBaDB), and I have a better understanding of the PG transaction system. This may be the reason why I could find this issue.

**Comment by ashwinstar** [comment]
This was fixed via ac34cd1eb5e081fed9a101f25961462df76287e9
